infrasim/dae.py

```diff
@@ -0,0 +1,628 @@
+import json
+import re
+
+"""
+  Note:
+  DAEProcessHelper is the entry to this module.
+  All other classes, including Const, Link, Enclosure, Drv, Expander and
+  DAEController,  are used to store internal data for building topology.
+
+"""
+
+
+class _Const:
+    DEFAULT_EXP_START_PHY = 8
+    NON_DEVICE = 0
+    END_DEVICE = 1
+    EXP_DEVICE = 2
+    HBA_DEVICE = 4
+    ACTIVE_DEVICE = 5
+    SES_DEVICE = 13
+    MAGIC_NUMBER = 0x23bc
+    BIN_VER = 0x0
+
+    @staticmethod
+    def check_range_conflict(phy0, number0, phy1, number1):
+        if not (phy0 >= (phy1 + number1) or (phy0 + number0) <= phy1):
+            return True
+        return False
+
+
+class _SNGenerator(object):
+    INDEX = 0
+
+    @classmethod
+    def getSN(cls):
+        sn = "Z4C03{0:03X}".format(cls.INDEX)
```

Comment:
❓ serial number matters between different vendors, there could be a more flexible implementation here. @chenc52 and @xiar has some insight.

Reply:
Agree.  Code here mainly take care about the requirements of, each drive has different SN within one node, and drive object in same position  (slot) should have same SN in different nodes.  
We may ask user to provide a prefix of SN in order to simulate different vendors here.